File: mobile/lib/internal/repository/product.dart
import 'package:dartz/dartz.dart';
import 'package:demo_store/internal/core/error/failures.dart';
import 'package:demo_store/internal/datasources/api.dart';
import 'package:demo_store/internal/entity/product/product.dart';
import 'package:demo_store/internal/entity/storage/storage.dart';

class ProductRepository {
  final DataSource source;

  ProductRepository(this.source);

  Future<Either<Failure, ProductList>> loadAll() async {
    final res = await source.get("/product/load_all");

    return res.fold((l) => Left(l), (r) => Right(ProductList.fromJson(r)));
  }

  Future<Either<Failure, void>> create(
      String name, double price, int stID) async {
    final res = await source
        .post("/product/create", {"name": name, "price": price, "st_id": stID});

    return res.fold((l) => Left(l), (r) => const Right(null));
  }

  Future<Either<Failure, void>> delete(int productID) async {
    final res = await source.post("/product/remove", {"product_id": productID});

    return res.fold((l) => Left(l), (r) => const Right(null));
  }

  Future<Either<Failure, StorageList>> loadStorageList() async {
    final res = await source.get("/product/storage_list");

    return res.fold((l) => Left(l), (r) => Right(StorageList.fromJson(r)));
  }
}

File: frontend/src/stores/product.ts
import { Failure } from "@/internal/entity/failure";
import { Product, Storage } from "@/internal/types/product";
import { defineStore } from "pinia";
import { useToast } from "vue-toastification";
import { siteAPI } from "./api";
import { err, ok, Result } from "neverthrow";

const toast = useToast();

interface loadAll {
  productList: Array<Product> | null;
}

interface loadProduct {
  product: Product | null;
}

interface create {
  product_id: number;
}

interface loadStorageList {
  storageList: Array<Storage> | null;
}

export const useProductStore = defineStore("product", {
  state: () => ({
    productList: null as Array<Product> | null,
    product: null as Product | null,
    storageList: null as Array<Storage> | null,
  }),
  getters: {},
  actions: {
    async loadAll() {
      const res = await siteAPI.get("/product/load_all");
      res.match(
        (r: loadAll) => {
          this.productList = r.productList;
        },
        (err: Failure) => {
          toast.error(err.message);
        }
      );
    },
    async create(name: string, price: number, st_id: number) {
      const res = await siteAPI.post("/product/create", {
        name: name,
        price: price,
        st_id: st_id,
      });
      return res.match(
        (r: create) => {
          toast.success(`Продукт №${r.product_id} успешно создан`);
          this.loadAll();
        },
        (e: Failure) => {
          toast.error(e.message);
        }
      );
    },
    async remove(productID: number) {
      const res = await siteAPI.post("/product/remove", {
        product_id: productID,
      });
      res.match(
        (_) => {
          toast.success(`Продукт №${productID} успешно удален`);
        },
        (err: Failure) => {
          toast.error(err.message);
        }
      );
    },
    async loadProduct(productID: number) {
      const res = await siteAPI.get(`/product/find/${productID}`);
      res.match(
        (r: loadProduct) => {
          this.product = r.product;
        },
        (err: Failure) => {
          toast.error(err.message);
        }
      );
    },
    async loadStorageList() {
      const res = await siteAPI.get(`/product/storage_list`);
      res.match(
        (r: loadStorageList) => {
          this.storageList = r.storageList;
        },
        (err: Failure) => {
          toast.error(err.message);
        }
      );
    },
  },
});
